In [1]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 17: No Such Thing as Too Much ---</h2><p>The elves bought too much eggnog again - <code>150</code> liters this time.  To fit it all into your refrigerator, you'll need to move it into smaller containers.  You take an inventory of the capacities of the available containers.</p>
<p>For example, suppose you have containers of size <code>20</code>, <code>15</code>, <code>10</code>, <code>5</code>, and <code>5</code> liters.  If you need to store <code>25</code> liters, there are four ways to do it:</p>
<ul>
<li><code>15</code> and <code>10</code></li>
<li><code>20</code> and <code>5</code> (the first <code>5</code>)</li>
<li><code>20</code> and <code>5</code> (the second <code>5</code>)</li>
<li><code>15</code>, <code>5</code>, and <code>5</code></li>
</ul>
<p>Filling all containers entirely, how many different <em>combinations of containers</em> can exactly fit all <code>150</code> liters of eggnog?</p>
</article>


In [2]:
from typing import Generator


containers = [20, 15, 10, 5, 5]
eggnog = 25


def combinations(eggnog: int, containers: list[int]) -> int:
    def dfs(i: int, eggnog: int, combination: list[int]) -> Generator[tuple[int, ...]]:
        if eggnog == 0:
            yield combination
        elif i < len(containers):
            yield from dfs(i + 1, eggnog - containers[i], combination + [containers[i]])
            yield from dfs(i + 1, eggnog, combination)

    yield from dfs(0, eggnog, [])


def different_combinations(eggnog, containers):
    return sum(1 for _ in combinations(eggnog, containers))


print(different_combinations(eggnog, containers))

4


In [3]:
containers = [11, 30, 47, 31, 32, 36, 3, 1, 5, 3, 32, 36, 15, 11, 46, 26, 28, 1, 19, 3]
eggnog = 150
print(different_combinations(eggnog, containers))

4372


<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 17: No Such Thing as Too Much ---</h2><p>The elves bought too much eggnog again - <code>150</code> liters this time.  To fit it all into your refrigerator, you'll need to move it into smaller containers.  You take an inventory of the capacities of the available containers.</p>
<p>For example, suppose you have containers of size <code>20</code>, <code>15</code>, <code>10</code>, <code>5</code>, and <code>5</code> liters.  If you need to store <code>25</code> liters, there are four ways to do it:</p>
<ul>
<li><code>15</code> and <code>10</code></li>
<li><code>20</code> and <code>5</code> (the first <code>5</code>)</li>
<li><code>20</code> and <code>5</code> (the second <code>5</code>)</li>
<li><code>15</code>, <code>5</code>, and <code>5</code></li>
</ul>
<p>Filling all containers entirely, how many different <em>combinations of containers</em> can exactly fit all <code>150</code> liters of eggnog?</p>
</article>


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>4372</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>While playing with all the containers in the kitchen, another load of eggnog <span title="Apparently, Amazon ships to the North Pole now.">arrives</span>!  The shipping and receiving department is requesting as many containers as you can spare.</p>
<p>Find the minimum number of containers that can exactly fit all <code>150</code> liters of eggnog.  <em>How many different ways</em> can you fill that number of containers and still hold exactly <code>150</code> litres?</p>
<p>In the example above, the minimum number of containers was two.  There were three ways to use that many containers, and so the answer there would be <code>3</code>.</p>
<p></p>
</article>

</main>


In [4]:
containers = [20, 15, 10, 5, 5]
eggnog = 25

min_length = len(min(combinations(eggnog, containers), key=len))

all_min_containers = [
    c for c in combinations(eggnog, containers) if len(c) == min_length
]


print(all_min_containers)
print("answer =", len(all_min_containers))

[[20, 5], [20, 5], [15, 10]]
answer = 3


In [5]:
containers = [11, 30, 47, 31, 32, 36, 3, 1, 5, 3, 32, 36, 15, 11, 46, 26, 28, 1, 19, 3]
eggnog = 150

min_length = len(min(combinations(eggnog, containers), key=len))

all_min_containers = [
    c for c in combinations(eggnog, containers) if len(c) == min_length
]


print(all_min_containers)
print("answer =", len(all_min_containers))

[[47, 31, 36, 36], [47, 31, 46, 26], [32, 36, 36, 46], [36, 32, 36, 46]]
answer = 4


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>4</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2015">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="17/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22No+Such+Thing+as+Too+Much%22+%2D+Day+17+%2D+Advent+of+Code+2015&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F17&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22No+Such+Thing+as+Too+Much%22+%2D+Day+17+%2D+Advent+of+Code+2015+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F17';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
